test/hotspot/jtreg/compiler/igvn/IntegerDivValueTests.java

```diff
@@ -0,0 +1,505 @@
+    }
+
+    @Run(test = {"testIntRange", "testIntRange2", "testIntRange3", "testIntRange4", "testIntRange5", "testIntRange6", "testIntRange7", "testIntRange8", "testIntRandomLimits"})
+    public void checkIntRanges(RunInfo info) {
+        for (int j = 0; j < 20; j++) {
+            int i1 = INTS.next();
+            int i2 = INTS.next();
+            checkInt(i1, i2);
+        }
+    }
+
+    @DontCompile
+    public void checkInt(int in, int in2) {
+        int a;
+        int b;
+        a = (in & 7) + 16;
+        Asserts.assertEquals(a / 12, testIntRange(in));
+
+        a = (in & 7) + 16;
+        Asserts.assertEquals(a / 4 > 3, testIntRange2(in));
+
+        a = (in & 31) + 16;
+        b = (in2 & 3) + 5;
+        Asserts.assertEquals(a / b > 4, testIntRange3(in, in2));
+
+        a = (in & 15);
+        b = (in2 & 3) + 1;
+        Asserts.assertEquals(a / b >= 0, testIntRange4(in, in2));
+
+        a = (in & 15) + 5;
+        b = (in2 & 3) + 1;
+        Asserts.assertEquals(a / b > 0, testIntRange5(in, in2));
+
+        a = (in & 15) + 5;
+        b = (in2 & 7) - 1;
+        Asserts.assertEquals(b == 0 ? false : a / b < -20, testIntRange6(in, in2));
+
+        a = (in & 15) + 5;
+        b = (in2 & 7) - 1;
+        Asserts.assertEquals(b == 0 ? false : a / b > 0, testIntRange7(in, in2));
+
+        a = (in & 31) + 128;
+        b = (in2 & 15) + 100;
+        Asserts.assertEquals(a / b, testIntRange8(in, in2));
+
+        int res;
+        try {
+            res = testIntRandomLimitsInterpreted(a, b);
+        } catch (ArithmeticException _) {
+            try {
+                testIntRandomLimits(a, b);
+                Asserts.fail("Expected ArithmeticException");
+                return; // unreachable
+            } catch (ArithmeticException _) {
+                return; // test succeeded, no result to assert
+            }
+        }
+        Asserts.assertEQ(res, testIntRandomLimits(a, b));
+    }
+
+    // Long variants
+
+    @ForceInline
+    private long getLongConstant(long value) {
+        // Simply return the given value to avoid javac already optimizing the operation away
+        return value;
+    }
+
+    private static final long LONG_CONST_1 = LONGS.next();
+    private static final long LONG_CONST_2 = LONGS.next();
+
+    @Test
+    //@IR(failOn = {IRNode.DIV_I, IRNode.DIV_L, IRNode.URSHIFT_I, IRNode.URSHIFT_L, IRNode.RSHIFT_I, IRNode.RSHIFT_L, IRNode.MUL_I, IRNode.MUL_L, IRNode.ADD_I, IRNode.ADD_L, IRNode.SUB_I, IRNode.SUB_L, IRNode.AND_I, IRNode.AND_L})
+    // This results in a series of nodes due to DivLNode::Ideal and in particular transform_long_divide, which operates on non-constant divisors.
+    // transform_long_divide splits up the division into multiple other nodes, such as MulHiLNode, which does not have a good Value() implemantion.
+    // When JDK-8366815 is fixed, these rules should be reenabled
+    // Alternatively, a better MulHiLNode::Value() implemantion should also lead to constant folding
```

Comment:
Could you have some temporary IR rule that now passes, but fails once `JDK-8366815` is fixed? Otherwise, I'm afraid we will miss these comments here, and they will never be cleaned up.

Reply:
Same elsewhere in this file.